frontend/src/main/scala/bloop/engine/Dag.scala

```diff
@@ -24,51 +25,59 @@ object Dag {
     val visiting = new scala.collection.mutable.LinkedHashSet[Project]()
     val dependents = new scala.collection.mutable.HashSet[Dag[Project]]()
     val projects = projectsMap.values.toList
-    def loop(project: Project, dependent: Boolean): Dag[Project] = {
+    val traces = new ListBuffer[RecursiveTrace]()
+    def loop(project: Project, dependent: Boolean): Option[Dag[Project]] = {
       def markVisited(dag: Dag[Project]): Dag[Project] = { visited.+=(project -> dag); dag }
-      def register(dag: Dag[Project]): Dag[Project] = {
-        if (dependent && !dependents.contains(dag)) dependents.+=(dag); dag
+      def register(dagOpt: Option[Dag[Project]]): Option[Dag[Project]] = {
+        dagOpt.map { dag =>
+          if (dependent && !dependents.contains(dag)) dependents.+=(dag); dag
+        }
       }
 
       register {
         // Check that there are no cycles
-        if (visiting.contains(project))
-          throw new RecursiveCycle(visiting.toList :+ project)
-        else visiting.+=(project)
-
-        val dag = visited.get(project).getOrElse {
-          markVisited {
-            project match {
-              case leaf if project.dependencies.isEmpty => Leaf(leaf)
-              case parent =>
-                val childrenNodes = project.dependencies.iterator.flatMap { name =>
-                  projectsMap.get(name) match {
-                    case Some(project) => List(project)
-                    case None =>
-                      val newMissingDeps = missingDeps.get(parent) match {
-                        case Some(prev) => name :: prev
-                        case None => name :: Nil
-                      }
-
-                      missingDeps.+=((parent, newMissingDeps))
-                      Nil
+        if (visiting.contains(project)) {
+          // We register the recursive cycle and return none
+          traces.+=(RecursiveTrace(visiting.toList :+ project))
+          None
+        } else {
+          visiting.+=(project)
+
+          val dag = visited.get(project).getOrElse {
+            markVisited {
+              project match {
```

Comment:
nit: the pattern match could be replaced with `if (project.dependencies.isEmpty) .. else ..`